chore: refactor invoice export link

hei-ryan opened the pull request.

hei-ryan committed:
chore: refactor invoice export link

nirinaHelena committed:
refactor: paginate result from service

nvagno commented on src/main/java/app/bpartners/api/service/InvoiceService.java:
```diff
@@ -73,78 +64,45 @@ public class InvoiceService {
+            userId, providedStatuses, providedArchiveStatus, filters, null, null);
+    int totalInvoice = invoices.size();
+    int nbPages = totalInvoice / MAX_SIZE;
+    for (int page = 1; page <= nbPages; page++) {
```
page from `0`

nvagno commented on 1d7ce579-46fb-483f-8bed-a9bbddb4ce85.pdf:
Put these files in a specific directory as testdata

nvagno commented on src/main/java/app/bpartners/api/service/InvoiceService.java:
```diff
@@ -73,78 +64,45 @@ public class InvoiceService {
-        .updatedAt(now)
+        .value(null)
+        .expirationDelay(null)
+        .updatedAt(Instant.now())
```
Use static import

nvagno commented on src/main/java/app/bpartners/api/service/event/InvoiceExportLinkRequestedService.java:
```diff
@@ -0,0 +1,125 @@
+            + providedTo
+            + " disponible";
+    // TODO must be the artisan email and admin as cc
+    // TODO var recipient = user.getDefaultHolder().getEmail();
```
implement these `TODO`

nirinaHelena committed:
test: InvoiceExportLinkRequestedServiceTest::generate_export_link_with_invoices_ok

nirinaHelena committed:
test: InvoiceServiceTest::generate_invoice_export_link

nirinaHelena committed:
refactor: use Mailer

nirinaHelena committed:
refactor: change invoiceExportLinkRequested.html to invoice_export_link_requested.html

nirinaHelena committed:
refactor: remove unused field DEFAULT_EXPIRATION_DELAY

nirinaHelena committed:
refactor: remove unused field DEFAULT_EXPIRATION_DELAY

nvagno commented on src/main/java/app/bpartners/api/service/InvoiceService.java:
```diff
@@ -73,78 +64,45 @@ public class InvoiceService {
+            userId, providedStatuses, providedArchiveStatus, filters, null, null);
+    int totalInvoice = invoices.size();
+    int nbPages = totalInvoice / MAX_SIZE;
+    for (int page = 0; page <= nbPages; page++) {
```
`<` but not `<=`

nvagno commented on src/main/java/app/bpartners/api/service/event/InvoiceExportLinkRequestedService.java:
```diff
@@ -0,0 +1,153 @@
+  private Context configureInvoiceLinkContext(
+      User user, LocalDate from, LocalDate to, String preSignedURL) {
+    Context context = new Context();
+    context.setVariable("userName", user.getDefaultHolder().getName());
```
`firstName` and `lastName` instead of `userName`

nvagno commented on src/main/resources/templates/invoice_export_link_requested.html:
```diff
@@ -0,0 +1,32 @@
+</head>
+<body>
+<p>
+    Bonjour,
```
Be more formal in the email

nirinaHelena committed:
refactor: make mail more formal

nirinaHelena committed:
test: InvoiceServiceTest::update_payment_status

nirinaHelena committed:
test: InvoiceServiceTest::duplicate_as_draft

nirinaHelena committed:
chore(invoiceService): update pagination logic for invoice export links

- Ensure the method returns 1 when total invoices are less than MAX_SIZE.
- Use Math.ceil to round up the number of pages for total invoices, ensuring correct pagination.

Dinasoa committed:
chore: use google geocode api instead of ban during address research

nirinaHelena committed:
refactor: use toList() instead of Collectors

hei-ryan commented on src/main/java/app/bpartners/api/service/InvoiceService.java:
```diff
@@ -73,78 +62,32 @@ public class InvoiceService {
-        .updatedAt(now)
+        .value(null)
+        .expirationDelay(null)
+        .updatedAt(Instant.now())
```
Import now() directly

hei-ryan commented on src/main/java/app/bpartners/api/service/event/InvoiceExportLinkRequestedService.java:
```diff
@@ -0,0 +1,167 @@
+            : providedTo.atStartOfDay();
+    var user = userRepository.getByIdAccount(accountId);
+    var userId = user.getId();
+    var from = fromLocalDatetime.atZone(ZoneId.systemDefault()).toInstant();
```
Set ZoneId to Europe/Paris ZoneId directly

hei-ryan commented on src/main/java/app/bpartners/api/service/event/InvoiceExportLinkRequestedService.java:
```diff
@@ -0,0 +1,167 @@
+    var user = userRepository.getByIdAccount(accountId);
+    var userId = user.getId();
+    var from = fromLocalDatetime.atZone(ZoneId.systemDefault()).toInstant();
+    var to = toLocalDateTime.atZone(ZoneId.systemDefault()).toInstant();
```
Set ZoneId to Europe/Paris ZoneId directly

hei-ryan commented on src/main/java/app/bpartners/api/service/event/InvoiceExportLinkRequestedService.java:
```diff
@@ -0,0 +1,167 @@
+        }
+      }
+      var invoicesZipFile = fileZipper.apply(List.of(invoicesFiles.toFile()));
+      zipFileId = randomUUID().toString();
```
Why the declaration of zipFileId and its value assignation is separated ?

nirinaHelena replied:
there are 2 assignments of zipFileId on line 94 and 96

hei-ryan commented on src/main/java/app/bpartners/api/service/event/InvoiceExportLinkRequestedService.java:
```diff
@@ -0,0 +1,167 @@
+      var invoicesZipFile = fileZipper.apply(List.of(invoicesFiles.toFile()));
+      zipFileId = randomUUID().toString();
+      s3Service.uploadFile(INVOICE_ZIP, zipFileId, userId, invoicesZipFile);
+      long expirationInSeconds = 3600L;
```
Set as static variable

hei-ryan commented on src/main/java/app/bpartners/api/service/event/InvoiceExportLinkRequestedService.java:
```diff
@@ -0,0 +1,167 @@
+      s3Service.uploadFile(INVOICE_ZIP, zipFileId, userId, invoicesZipFile);
+      long expirationInSeconds = 3600L;
+      var preSignedURL = s3Service.presignURL(INVOICE_ZIP, zipFileId, userId, expirationInSeconds);
+      htmlBody =
```
This htmlBody value assignement is override in line 104 anyway ...

hei-ryan commented on src/main/java/app/bpartners/api/repository/jpa/InvoiceJpaRepository.java:
```diff
@@ -26,6 +27,8 @@ List<HInvoice> findAllByIdUserAndArchiveStatusAndTitleContainingIgnoreCase(
 
   List<HInvoice> findByIdUserAndRef(String idAccount, String ref);
 
+  int countByIdUserAndCreatedDatetimeBetween(String idUser, Instant from, Instant to);
```
No ! I already said it's not the createdDatetime that should be filtered, but sendingDate

hei-ryan commented on src/main/java/app/bpartners/api/endpoint/event/model/InvoiceExportLinkRequested.java:
```diff
@@ -0,0 +1,37 @@
+@Builder
+@Data
+@ToString
+public class InvoiceExportLinkRequested extends PojaEvent {
```
I think this event must be handle with EVENT_STACK_2 because it needs some memory support ... 

Just override getEventStack() { return EVENT_STACK_2;}
